packages/compass-aggregations/src/utils/pipeline-storage.ts

```diff
@@ -57,88 +41,109 @@ function savedPipelineToText(pipeline: StoredPipeline['pipeline']): string {
+        })
+      )
+      .optional()
+      .describe('Legacy property to store pipeline. Use pipelineText.'),
+    host: z.string().nullable().optional(),
+    pipelineText: z.string().optional(),
+    lastModified: z
+      .number()
+      .transform((x) => new Date(x))
+      .optional(),
+  })
+  .transform((input) => {
+    const {
+      pipeline: legacyPipelineArray,
+      pipelineText,
+      ...restOfInput
+    } = input;
+    return {
+      ...restOfInput,
+      pipelineText: pipelineText ?? savedPipelineToText(legacyPipelineArray),
+    };
+  });
+
+export type StoredPipeline = z.output<typeof PipelineSchema>;
+
+type SaveablePipeline = Omit<z.input<typeof PipelineSchema>, 'pipeline'>;
 
 export class PipelineStorage {
+  private readonly userData: UserData<typeof PipelineSchema>;
+  constructor(basePath?: string) {
+    this.userData = new UserData(PipelineSchema, {
+      subdir: 'SavedPipelines',
+      basePath,
+    });
+  }
+
+  private mergeStats(pipeline: StoredPipeline, stats: Stats): StoredPipeline {
+    return {
+      ...pipeline,
+      lastModified: new Date(stats.ctimeMs),
+    };
+  }
+
   async loadAll(): Promise<StoredPipeline[]> {
     try {
-      const dir = getDirectory();
-      const files = (await fs.readdir(dir))
-        .filter((file) => file.endsWith('.json'))
-        .map((file) => path.join(dir, file));
-
-      return (
-        await Promise.all(files.map((filePath) => this._loadOne(filePath)))
-      ).filter(Boolean) as StoredPipeline[];
+      const { data } = await this.userData.readAllWithStats({
+        ignoreErrors: false,
+      });
+      return data.map(([item, stats]) => this.mergeStats(item, stats));
     } catch (e) {
+      debug('Failed to read saved pipelines.', e);
       return [];
     }
   }
 
-  async load(id: string): Promise<StoredPipeline | null> {
-    return this._loadOne(path.join(getDirectory(), `${id}.json`));
+  private async loadOne(id: string): Promise<StoredPipeline> {
+    const [item, stats] = await this.userData.readOneWithStats(id);
+    return this.mergeStats(item, stats);
   }
 
-  async _loadOne(filePath: string): Promise<StoredPipeline | null> {
-    try {
-      const [data, stats] = await Promise.all([
-        this._getFileData(filePath),
-        fs.stat(filePath),
-      ]);
-      if (!hasAllRequiredKeys(data)) {
-        return null;
-      }
-      return {
-        ...data,
-        lastModified: stats.mtimeMs,
-        pipelineText: data.pipelineText ?? savedPipelineToText(data.pipeline),
-      };
-    } catch (err) {
-      debug(`Failed to load pipeline ${path.basename(filePath)}`, err);
-      return null;
-    }
+  async createOrUpdate(id: string, attributes: SaveablePipeline) {
+    const pipelineExists = Boolean(
+      await this.userData.readOne(id, {
+        ignoreErrors: true,
+      })
+    );
+    return await (pipelineExists
+      ? this.updateAttributes(id, attributes)
+      : this.create(attributes));
```

Comment:
We're adding another read on top of all the reads that `update` will be doing, I think we need to reconsider some thing about internal workings of user data based on the use-cases we have in compass, doesn't look like the way it works fits well to what we need to do most of the time